c2rust-analyze/src/rewrite/expr/hir_op.rs

```diff
@@ -244,47 +263,112 @@ impl<'a, 'tcx> intravisit::Visitor<'tcx> for HirRewriteVisitor<'a, 'tcx> {
+                    let elem = Rewrite::Index(Box::new(arr), Box::new(idx));
+                    Rewrite::Ref(Box::new(elem), mutbl_from_bool(*mutbl))
+                }
+
+                mir_op::RewriteKind::MutToImm => {
+                    // `p` -> `&*p`
+                    let place = Rewrite::Deref(Box::new(hir_rw));
+                    Rewrite::Ref(Box::new(place), hir::Mutability::Not)
+                }
+
+                mir_op::RewriteKind::RemoveAsPtr => {
+                    // `slice.as_ptr()` -> `slice`
+                    assert!(matches!(hir_rw, Rewrite::Identity));
+                    self.get_subexpr(ex, 0)
+                }
+
+                mir_op::RewriteKind::RawToRef { mutbl } => {
+                    // &raw _ to &_ or &raw mut _ to &mut _
+                    Rewrite::Ref(Box::new(self.get_subexpr(ex, 0)), mutbl_from_bool(*mutbl))
+                }
+
+                mir_op::RewriteKind::CellNew => {
+                    // `x` to `Cell::new(x)`
+                    Rewrite::Call("std::cell::Cell::new".to_string(), vec![Rewrite::Identity])
+                }
+
+                mir_op::RewriteKind::CellGet => {
+                    // `*x` to `Cell::get(x)`
+                    Rewrite::MethodCall(
+                        "get".to_string(),
+                        Box::new(self.get_subexpr(ex, 0)),
+                        vec![],
+                    )
+                }
+
+                mir_op::RewriteKind::CellSet => {
+                    // `*x` to `Cell::set(x)`
+                    let deref_lhs = assert_matches!(ex.kind, ExprKind::Assign(lhs, ..) => lhs);
+                    let lhs = self.get_subexpr(deref_lhs, 0);
+                    let rhs = self.get_subexpr(ex, 1);
+                    Rewrite::MethodCall("set".to_string(), Box::new(lhs), vec![rhs])
+                }
             }
+        };
+
+        let is_addr_of_expansion = || {
+            let rvalues: Vec<_> = locs
+                .into_iter()
+                .map(|loc| {
+                    self.mir
+                        .stmt_at(loc)
+                        .left()
+                        .and_then(|s| s.kind.as_assign().cloned().map(|(_, rv)| rv))
+                })
+                .collect();
+            use mir::Rvalue;
+            matches!(
+                rvalues[..],
+                [Some(Rvalue::Ref(..)), Some(Rvalue::AddressOf(..))]
+            )
+        };
+
+        use mir_op::RewriteKind::*;
+        if !all_rws_unflattened.is_empty() {
+            hir_rw = match &all_rws_unflattened[..] {
+                [[rw @ RawToRef { .. }]] if callsite_span != ex.span && is_addr_of_expansion() => {
+                    rewrite_from_mir_rws(rw, hir_rw)
+                }
+                [rws @ [OffsetSlice { .. }, SliceFirst { .. }]] => rws
+                    .iter()
+                    .fold(hir_rw, |acc, rw| rewrite_from_mir_rws(rw, acc)),
+                [[rw]] => rewrite_from_mir_rws(rw, hir_rw),
```

Comment:
With the filter removal from the previous comment, I think these two cases could be combined into a single case that applies when there is only one location.

Reply:
I've combined them in 849d4d16 without the filter

Reply:
I'm not sure I like the https://github.com/immunant/c2rust/commit/849d4d163316880197fadf100160613985a5e5ca approach, though. I think the original constrains the space well-enough that we can catch any obviously-weird cases in the panic

Reply:
I've added back the constraints and kept it one match arm in fbe4e2d0

Reply:
It works either way, but the `[rws] => ...` version is what I was thinking.  I think it's more important to filter based on the number of locations than on the number of rewrites.  If there's only one location but it has many rewrites, the most obvious thing to do is to apply them all in order (and `mir_op` is careful to generate rewrites in a sensible order to allow this).  It's only when there are multiple locations that it becomes unclear what exactly you should do with each set of rewrites.

Reply:
Agreed, addressed in 8114679a